src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -6,58 +6,126 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
-using System.Diagnostics.CodeAnalysis;
 using System.Runtime.CompilerServices;
+using System.Runtime.Serialization;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Serialization;
 
-internal sealed class SolutionStateChecksums : IChecksummedObject
+[DataContract]
+internal readonly struct AssetHint
 {
-    public Checksum Checksum { get; }
-
-    public Checksum Attributes { get; }
-    public ImmutableArray<ProjectId> ProjectIds { get; }
-    public ChecksumCollection Projects { get; }
-    public ChecksumCollection AnalyzerReferences { get; }
-    public Checksum FrozenSourceGeneratedDocumentIdentity { get; }
-    public Checksum FrozenSourceGeneratedDocumentText { get; }
-
-    public SolutionStateChecksums(
-        Checksum attributes,
-        ImmutableArray<ProjectId> projectIds,
-        ChecksumCollection projects,
-        ChecksumCollection analyzerReferences,
-        Checksum frozenSourceGeneratedDocumentIdentity,
-        Checksum frozenSourceGeneratedDocumentText)
+    public static readonly AssetHint None = default;
+
+    [DataMember(Order = 0)]
+    public readonly ProjectId? ProjectId;
+    [DataMember(Order = 1)]
+    public readonly DocumentId? DocumentId;
+
+    private AssetHint(ProjectId? projectId, DocumentId? documentId)
     {
-        Contract.ThrowIfFalse(projectIds.Length == projects.Children.Length);
+        ProjectId = projectId;
+        DocumentId = documentId;
+    }
+
+    public static implicit operator AssetHint(ProjectId projectId) => new(projectId, documentId: null);
+    public static implicit operator AssetHint(DocumentId documentId) => new(documentId.ProjectId, documentId);
+}
+
+internal readonly struct ChecksumsAndIds<TId> where TId : IObjectWritable
```

Comment:
encapsulates a list of document or project ids, and the checksums for their corresponding objects.